blueocean-dashboard/src/main/js/components/karaoke/components/LogConsole.jsx

```diff
@@ -180,7 +185,8 @@ export class LogConsole extends Component {
                     </a>
                 </div>}
                 {!isLoading && lines.map((line, index) => (
-                    <LogLine prefix={prefix}
+                    <LogLine key={index}
```

Comment:
I'm confused about the `key` usage vs. `PureComponent`. Also using indices for keys isn't particularly useful - the line text would be more appropriate.

Reply:
> I'm confused about the `key` usage vs. `PureComponent`.

PureComponent is the main fix of this PR, however I noticed that there was a `key` attribute on the `p` element of `LogLine`, which will have no effect on react's rendering.

> using indices for keys isn't particularly useful - the line text would be more appropriate.

Definitely! Generally indexes for keys are not valid and are considered an anti-pattern. However since this array is never ordered, or filtered, and is monotonically increasing, the index is the best key here. I decided to define the key explicitly here, rather than rely on a default key provided by React.

With regards to using the line text instead of the index for the key, there may be a problem with that - for instance a user may have the line "hello world" twice in their logs, which could cause issues

Let me know if i've misunderstood anything :+1:

Reply:
Good point about the line as a key. But why have a key at all? Is it required for the `PureComponent`?

Reply:
@kzantow Keys are only needed when react elements are contained within arrays, and are unrelated to Pure components 

The react docs are great in explaining this in more detail than I possibly could! Unintentionally the react docs describe this exact scenario too:

https://reactjs.org/docs/lists-and-keys.html#extracting-components-with-keys

Reply:
Right, I understand why to use keys. Question is... are they actually necessary with the `PureRender` component? They are to find the original components, I'm guessing. So good. :)